### DISCUSSION TITLE
What hardware is available for a snapclient?

### DISCUSSION THREAD

**Discussion by IRQcko** [snapcast/snapcast#1055] [Q&A] [answered]
Given current RPi shortage I am trying to figure out what cheap but stable hardware to use for deploying multiple snapclients. I was trying to use ESP32 but without much success. What else is suitable **and** **available** for snapclients?

**Comment by dgalli1** [comment]
Best I found are the khadas boards, but they come out a little to price to just play audio. 
A Zero with a hifiberry still seems like the better combination.

Another thing that im currently eying is using something like this:
https://www.arylic.com/products/up2stream-pro-receiver-board
And trying to code a proxy for snapcast client to pass the audio to this. But i as I haven't even orderd one I wouldn't hold my breath.

**Reply by IRQcko** [reply]
Raspberry Pi Zero is really good option, but not available in mid-term future.
Possible solution could be OrangePi, but OPi ecosystem is a mess. Zero boards are only 25-pin (so lacking some I2S pins), OPi One has 40-pin header, but lacks wifi.

**Reply by dgalli1** [reply]
yeah it's a mess i also thought about using a orang pi with a wifi adapter, but i really don't want to add another failure point.

Currently, I'm really leaning into buying something with airplay support and then just writing a custom client which proxies the whole audio stream via airplay or dlna.

**Reply by dgalli1** [reply]
Nvm just looked into it a bit more, this seems to be even more feasible than I thought. 

Those arylic boards seem to be using the LinkPlayAPI.

You can basically just send them an url with a stream.
So the proxy client could basically just translate the snapcast socket commands and then send them to the client software without any big issues. https://github.com/MClarkDev/ArylicHTTP/blob/master/src/ArylicHTTP.cpp

I just ordered one of their boards and will report back once it arrived and i got to developing this software.

**Reply by kingosticks** [reply]
These boards are interesting in general because they seem to have official Spotify Connect support, rather than librespot-based. I didn't expect that. Turns out it's implemented in this "Linkplay A31" thing (chip??) that's in a few other products, such as [wiim](https://www.wiimhome.com/Overview).

But I don't follow how your system would work. Where would you be transforming the chunked snapcast audio data into a "normal" file-like stream for this board to play? And you would be undoing all the synchronisation aspects of snapcast so is snapcast even the right fit for this setup? I might be misunderstanding here, apologies.

**Reply by dgalli1** [reply]
Yeah, that's the issue synchronization will not be doable but for my usecase it shouldn't be too big of an issue as all the speakers are in separate rooms, I will probably try to get to it via static delays, which is of course nowhere near as good as a normal snapcast client.

But considering that I can get an audio signal from those things for 70.- instead of somehow getting a pi and hifiberry or similar makes the downside worth it for me.

**Reply by kingosticks** [reply]
There are 2nd hand / refurb mini PCs on ebay for under £60. Would be much simpler than trying to bridge two different systems.
* [ "HP Mini PC, AMD 1.2GHz Dual Core, 4GB RAM, 8GB SSD, Ubuntu Linux + PSU" for £33! ](https://www.ebay.co.uk/itm/114847452861?mkcid=16&mkevt=1&mkrid=711-127632-2357-0&ssspo=ma1XvSCNTie&sssrc=2047675&ssuid=cK42yhSZTva&var=414997658107&widget_ver=artemis&media=COPY)
* https://www.ebay.co.uk/itm/165691849552?mkcid=16&mkevt=1&mkrid=711-127632-2357-0&ssspo=mQWoa1ovTle&sssrc=2047675&ssuid=cK42yhSZTva&widget_ver=artemis&media=COPY
* https://www.ebay.co.uk/itm/114847452861?mkcid=16&mkevt=1&mkrid=711-127632-2357-0&ssspo=ma1XvSCNTie&sssrc=2047675&ssuid=cK42yhSZTva&var=414997658107&widget_ver=artemis&media=COPY (needs RAM and storage but still works out much cheaper than Pi)

**Reply by dgalli1** [reply]
Yes that would also be an option, although I'm not a huge fan of having such devices running 24/7 fanless boards which use about 5 watts seem like the better option to me.

**Reply by kingosticks** [reply]
I appreciate that. Depends where you want to spend the resources. Making it work and keeping or working, or your electricity bill. I get the idea why the SBCs are so good but the ThinkCentre Tiny stuff is ~10W, you won't hear it and you won't notice it on your bills. Best of luck with it, do keep us updated.

**Comment by kantrol** [comment]
Just finished the configuration of some  [TL WR-710N](https://openwrt.org/toh/tp-link/tl-wr710n) travel routers. Flashed them with Openwrt and installed  [Openwrt snapclient](https://github.com/badaix/snapos/blob/develop/openwrt/README.md)
 
Works brilliant! As they are integrated in the power wall plug, you only need to attach a cheap USB sound card, very compact setup.
I was lucky to buy one TL WR-710N  for 6€, but usually you get them for 10-15€.
 
As the OpenWRT page mentiones, an 8MB Flash version should be selected (not the 2.0). As the WR-710 seems not to run well with the latest Openwrt version, I build the system with version 19.07.10.

**Reply by bw118** [reply]
Hello, I've been trying to do that but can't for the life of me do it...

https://github.com/badaix/snapos/issues/24

Can you help? TY!

**Reply by CarlosDerSeher** [reply]
Did you try to change PKG_VERSION and PKG_SOURCE_VERSION in https://github.com/badaix/snapos/blob/master/openwrt/snapcast/Makefile to a more recent value, v0.26.0 is the current version. Probably the problem is already fixed there

**Reply by bw118** [reply]
Thank you for the answer! I'll try this when I find some time... Life is too crazy atm... :-/

**Comment by CarlosDerSeher** [comment, marked as answer]
> I was trying to use ESP32 but without much success.

Did you try my develop branch?
https://github.com/CarlosDerSeher/snapclient

I am still debugging but it is very usable with my lyrat v4.3 board. Some other codecs are also supported and it shouldn't be to hard to add others.

**Reply by CarlosDerSeher** [reply]
So it's been a while and the project is mostly stable, at least if using FLAC. I use it daily and it works very very good.

**Reply by IRQcko** [reply]
Thank You for Your work. I will dive deep into it again when I get little more time.

**Reply by zierbeek** [reply]
oh well, I have quite a few speakers in different places which should get wireless audio. The Lyrat is rather 'expensive', for what I find. I have quite a lot of normal esp32's on hand.

**Reply by luar123** [reply]
You can use a normal ESP32, just get some simple DACs like PCM5102a.

**Reply by CarlosDerSeher** [reply]
Yes a "normal" esp32 without psram will work. The only restriction is a low enough buffer time around 700 to 800ms. I designed a custom PCB with the codec adau1961 and a "normal" esp32 and the performance of the esp32  is excellent. Didn't have any noticable glitches in audio so far :)

**Reply by zierbeek** [reply]
could you please explain your project a bit? I'm unable to see what I should order or follow to get that started :)

**Reply by CarlosDerSeher** [reply]
I know I have to update the Readme about that.

For now you could check out https://github.com/CarlosDerSeher/snapclient/discussions/29 maybe this will already put you on track.

**Comment by dgalli1** [comment]
I got a new favorite
NanoPi NEO
entry level dac: https://de.aliexpress.com/item/1005003301431394.html
Digital Output: https://hwgroup-bg.com/nanohat-spdif/

Means one streamer will cost you about 30$ and it's tiny.
If you need wifi you can grab a nanopi neo air
